### ISSUE TITLE
[🐛] PhoneAuthListener: successCb and errorCb are not called, e2e test false-positive

### ISSUE THREAD

**Issue by Avishayy** [invertase/react-native-firebase#5917]
## Issue

I've been migrating my code base from 0.5 to 0.6 along with updating react-native from 0.61 to 0.66, one of the issues that arose is `PhoneAuthListener.on()` not calling its `successCb`.

Take a look at [this snippet](https://github.com/invertase/react-native-firebase/blob/6a1125cad1bb79a169ca073cb7ddc12b467cc7ef/packages/auth/lib/PhoneAuthListener.js#L208-L220):
```
    if (isFunction(errorCb)) {
      const subscription = this._auth.emitter.addListener(this._publicEvents.error, () => {
        errorCb;
        subscription.remove();
      });
    }

    if (isFunction(successCb)) {
      const subscription = this._auth.emitter.addListener(this._publicEvents.success, () => {
        successCb;
        subscription.remove();
      });
    }
```

Notice that `errorCb` and `successCb` are never called, I wanted to fix it by myself but then I realised they need to receive `phoneAuthError` and `PhoneAuthSnapshot` respectively. I'm not proficient enough in the library to understand what's the right way to do it so I'm reporting it here.

At the meantime I resorted to using `then` on `PhoneAuthListener`.

---

I removed the rest of the template because it is irrelevant.

**Comment by mikehardy** [comment]
I believe the understanding of the code is incorrect. I just read through and thought about it, and in plain english here is my understanding of control flow:

1) if success or error callback is a function, built a subscription closure that calls them then removes itself.
2) add that subscription closure (which is still live, with instructions to call the callback, then remove itself) as listeners for the events

Now when the event fires, it will call your callback once, then remove itself.

it's a form of one-shot event listener subscription

it was a port of the previous react-native API underlying, which has since been removed. It's a tested pattern :thinking: 

My commit: https://github.com/invertase/react-native-firebase/commit/5eb2f599e93ccecd91c800018959f9dc370f1e24

I linked the reference in that commit message

I do not believe this test would pass if the successCb wasn't called:

https://github.com/invertase/react-native-firebase/blob/6a1125cad1bb79a169ca073cb7ddc12b467cc7ef/packages/auth/e2e/phone.e2e.js#L118-L130

Given we have an e2e test suite that probes this with success, I'm going to close this as the burden of proof is really high. Happy to reopen but criteria are a modification of the e2e suite that shows the e2e suite is in error or the underlying code fails in a way you can show with an e2e test ?

**Comment by Avishayy** [comment]
> if success or error callback is a function, built a subscription closure that **calls them** then removes itself.  

That's what you say, but the code does something completely else? Either I'm completely illiterate or JavaScript has a feature that calls empty statements with functions name along with magically getting the right parameters to them?

Quoting the snippet I linked, again:
https://github.com/invertase/react-native-firebase/blob/6a1125cad1bb79a169ca073cb7ddc12b467cc7ef/packages/auth/lib/PhoneAuthListener.js#L208-L220

Yes, if success or error callbacks are a function, it builds a subscription closure that **uses the function name as a statement**, i.e. no parentheses, like so:
https://github.com/invertase/react-native-firebase/blob/6a1125cad1bb79a169ca073cb7ddc12b467cc7ef/packages/auth/lib/PhoneAuthListener.js#L210
https://github.com/invertase/react-native-firebase/blob/6a1125cad1bb79a169ca073cb7ddc12b467cc7ef/packages/auth/lib/PhoneAuthListener.js#L217

The functions are never called! I looked already at the commit you referenced an it __did__ call the functions by passing them to `once`:
```js
this._auth.emitter.once(this._publicEvents.error, errorCb);
```
and by looking at the reference in the commit, the implementation of `once` was:
```js
  once(
    eventType: string,
    listener: Function,
    context: ?Object,
  ): EmitterSubscription {
    return this.addListener(eventType, (...args) => {
      this.removeCurrentListener();
      listener.apply(context, args);
    });
  }
```
and the critical part is
```js
      listener.apply(context, args);
```
which actually **calls**  the function,  unlike the current implementation.

I don't know why the e2e test is broken but the only reason I am here is because I had perfectly working code that used `successCb` and that code got broken, I eventually found the issue and reported here.

**Comment by mikehardy** [comment]
Interesting for sure. Easy to settle it. How's it go when you change those two lines in listener construction to the call you expect with parentheses and args directly in node modules?

If it works we have a false positive test and a bug, this gets reopened, you can make a PR and use patch-package to get moving locally and the tests need a good scrub

**Comment by Avishayy** [comment]
I did try it, the function gets called, but I don't know how to get the right args to pass to these functions, as I said - I'm not proficient in this library at all.

I looked at the test again and I know why it's a false-positive, nothing actually checks that `successCb` was called, the fact it returns a resolve promise is irrelevant because its return value is never used.

**Comment by Avishayy** [comment]
`on()` doesn't even return a Promise, so `await`ing on it doesn't do anything either.

**Comment by mikehardy** [comment]
Ah, that's bogus then (the test false positive). Seems clear that everyone is using onAuthStateChanged or onUserChanged listener in their flows ;-), I mean to say: everyone is successfully using phone auth (I have certainly) but I guess the use case is more "fire and forget" on the call, then listen for state change events on the other handlers. Otherwise someone surely would have reported it before now with the code busted like that. That was in March!

Upstream once code was:

```javascript
  once(
    eventType: string,
    listener: Function,
    context: ?Object,
  ): EmitterSubscription {
    return this.addListener(eventType, (...args) => {
      this.removeCurrentListener();
      listener.apply(context, args);
    });
  }
```

So the code here should be

```javascript
      const subscription = this._auth.emitter.addListener(this._publicEvents.error, (...args) => {
        subscription.remove();
        errorCb.apply(args);
      });
```

untested, but could work? this is reopen now for test scrubbing and fix. Something *like* the above could work though?

**Comment by mikehardy** [comment]
Got a branch going here where I've scrubbed the e2e tests and can correctly demonstrate the current implementation is broken as you rightly brought up @Avishayy 

I've got the success callback fixed (and demonstrated working with tests), and I've altered the failure callback in the same pattern but it's not happy with the tests yet, and 90% of a repair is no repair.

Might take a little while to land it, but it is in progress https://github.com/invertase/react-native-firebase/commits/%40mikehardy/phoneauth-issue5917

**Comment by mikehardy** [comment]
@Avishayy I'll need your help on this one. I'm preparing a new release and #5933 will be a part of it I hope. I'd like you to grab the patches.zip off this action on the associated PR and see if your code is working again with the PR integrated via patch-package:

https://github.com/invertase/react-native-firebase/actions/runs/1575894049
https://github.com/ds300/patch-package

Near as I can tell, since the code has been definitely not working since March, you may be literally the only person using it, thus without you testing it I won't have any second opinion, and I'll default to merging+releasing, but it still may not work? So please give it a check if you can

Thanks!

**Comment by Avishayy** [comment]
I'll check it soon, thank you for working on it!

**Comment by mikehardy** [comment]
I just cleared out the queue of PRs waiting to be merged, and did some releases - the only thing outstanding now is this phone callback PR, so the patch set is now much cleaner if you grab it from here for testing

https://github.com/invertase/react-native-firebase/actions/runs/1579792044

**Comment by Avishayy** [comment]
I can confirm it works, thanks @mikehardy!

**Comment by mikehardy** [comment]
Yes! Thanks @Avishayy - I'll merge this and release it shortly, there will be no change from your patch set to the released version and the thing I love about patch-package is it gives you a nice warning in the terminal when the release comes out and you need to pay attention again (to update the patch, or remove it...) so you should be unblocked already while the release may take a few days. Cheers
